### ISSUE TITLE
Get access to custom classes via raw YOLOE output

### ISSUE THREAD

**Issue by samb271** [ultralytics/ultralytics#22279]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi! Thank you for your work and constant support. 

I am trying to run YOLOE using Triton Inference Server client, specifically the gRPC API. At export, I set a few custom classes that I would like the model to detect at inference. However, **_I seem unable to extract the predictions for those custom classes from the resulting outputs._** gRPC outputs raw logit tensors, and I'm quite confused about how to debug this. I'm probably doing something very wrong.

Here's some code snippets to give context.

I first export the ONNX model using the following script:

```
# Based on https://docs.ultralytics.com/guides/triton-inference-server/

from ultralytics import YOLOE
from pathlib import Path
from orchestrator.constants import YOLOE_CLASSES

model = YOLOE("yoloe-11l-seg.pt", task="segment")

metadata = []


def export_cb(exporter):
    metadata.append(exporter.metadata)


model.add_callback("on_export_end", export_cb)

names = YOLOE_CLASSES
model.set_classes(names, model.get_text_pe(names))

export_model = model.export(format="onnx", dynamic=True, nms=True)

# Convert metadata dict → Python-safe string (not JSON!)
# Ultralytics Triton backend calls eval(), so this must look like a Python dict
metadata_str = str(metadata[0])

model_name = "yoloe"
triton_repo_path = Path("tmp")
triton_model_path = triton_repo_path / model_name
(triton_model_path / "1").mkdir(parents=True, exist_ok=True)

Path(export_model).rename(triton_model_path / "1" / "model.onnx")

config = f"""
# Add metadata
parameters {{
  key: "metadata"
  value {{
    string_value: "{metadata_str.replace('"', '\\"')}"
  }}
}}

platform: "onnxruntime_onnx"

max_batch_size: 0

# (Optional) TensorRT acceleration
optimization {{
  execution_accelerators {{
    gpu_execution_accelerator {{
      name: "tensorrt"
      parameters {{
        key: "precision_mode"
        value: "FP16"
      }}
      parameters {{
        key: "max_workspace_size_bytes"
        value: "3221225472"
      }}
      parameters {{
        key: "trt_engine_cache_enable"
        value: "1"
      }}
      parameters {{
        key: "trt_engine_cache_path"
        value: "/models/{model_name}/1"
      }}
    }}
  }}
}}
"""

with open(triton_model_path / "config.pbtxt", "w") as f:
    f.write(config)

print("Export complete. Triton repo ready at:", triton_repo_path)
```

The exported configuration file reflects the custom classes:

`"...'task': 'segment', 'batch': 1, 'imgsz': [640, 640], 'names': {0: 'person', 1: 'car', 2: 'truck', 3: 'excavator', 4: 'hard hat', 5: 'safety vest'}, 'args': {'batch': 1, 'half': False, 'dynamic': True, 'simplify': True, 'opset': None, 'nms': True}, 'channels': 3}"`

I can successfully get detections using these custom classes using the HTTP client if I instantiate a YOLOE class like such:

```
from tritonclient.http import InferenceServerClient

triton_client = InferenceServerClient(url=TRITON_URL)
yoloe = YOLOE("http://triton:8000/yoloe", task="segment")

results = yoloe.predict(image)
boxes = results[0].boxes
class_ids = boxes.cls.cpu().numpy().astype(int)
confs = boxes.conf.cpu().numpy()
xyxys = boxes.xyxy.cpu().numpy()
```

However I wish to proceed with the gRPC API, which from my understanding cannot be used to directly instantiate a YOLOE class and must instead be used like this:

```
from tritonclient.grpc import InferenceServerClient
TRITON_URL = os.getenv("TRITON_URL", "localhost:8001")

client = InferenceServerClient(TRITON_URL)

inp = InferInput("images", tensor.shape, "FP32")
inp.set_data_from_numpy(tensor)

out0 = InferRequestedOutput("output0")
out1 = InferRequestedOutput("output1")

res = client.infer(MODEL_NAME, inputs=[inp], outputs=[out0, out1])

output0 = res.as_numpy("output0")
output1 = res.as_numpy("output1")

print(f"Frame {frame_count} processed. Output shapes:", output0.shape, output1.shape)
```

The output of this is the raw logits:

`Frame 0 processed. Output shapes: (1, 42, 8400) (1, 32, 160, 160)`

I extract the class IDs with this code:

```
def decode_yoloe_output(output0, conf_thresh=0.5, nms_thresh=0.2):

    batch_detections = []

    output0 = np.transpose(output0, (0, 2, 1))  # (N, 8400, 42)

    for b in range(output0.shape[0]):
        pred = output0[b]  # shape (8400, 42)
        boxes, confidences, class_ids = [], [], []

        for i in range(len(pred)):
            cx, cy, w, h = pred[i, 0:4]
            obj_conf = 1 / (1 + np.exp(-pred[i, 4]))
            class_logits = pred[i, 5:11]  # 6 custom classes
            class_scores = 1 / (1 + np.exp(-class_logits))
            cls_id = int(np.argmax(class_scores))
            score = class_scores[cls_id] * obj_conf
```

However, the class score for the first 6 classes (the custom classes) are nonsensical:

`[0.50000054 0.50000024 0.5000038  0.5        0.5000057  0.10298181]`

Do I need to be looking at the first 6 classification logits to get my custom classes? Any help would be greatly appreciated! 

Thank you

**Comment by UltralyticsAssistant** [comment]
👋 Hello @samb271, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get support faster—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Given your Triton gRPC/YOLOE output question, to help us reproduce and investigate, please include:
- Exact Ultralytics version (`ultralytics.__version__`) and Triton/ONNX Runtime versions
- The precise export command/args you used and the ONNX opset if set
- A small sample input image and the corresponding raw output tensors (shapes and a few values)
- Your Triton `config.pbtxt` (including metadata) and the output tensor names requested via gRPC
- The preprocessing steps used to create `images` input and any postprocessing you apply to the raw outputs
- Confirmation that HTTP works with the same model/inputs and what differs with gRPC
- A compact script/notebook that runs end-to-end on a single image as an [MRE](https://docs.ultralytics.com/help/minimum-reproducible-example/) 🙏

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
> `yoloe = YOLOE("http://triton:8000/yoloe", task="segment")`

You can just replace `http` with `grpc` to use gRPC. And change the port number to the gRPC port. Default is 8001.

**Comment by samb271** [comment]
Hi, thank you for the quick answer!

If I simply load the model like this:

```
try:
    yoloe = YOLOE("grpc://triton:8001/yoloe", task="segment")
    logging.info(f"Loaded YOLOE model")
except Exception as e:
    logging.error(f"Failed to load YOLOE: {e}")
    exit(1)
```

And then try an inference call as such:

```
results = yoloe.predict(inp)
```

I get the following error:

`Ultralytics 8.3.181 🚀 Python-3.12.11 torch-2.8.0+cu128 CUDA:0 (NVIDIA GeForce RTX 3090, 24253MiB)
c19:36:34 [ERROR] Inference failed: model='None' is not a supported model format. Ultralytics supports: ('PyTorch', 'TorchScript', 'ONNX', 'OpenVINO', 'TensorRT', 'CoreML', 'TensorFlow SavedModel', 'TensorFlow GraphDef', 'TensorFlow Lite', 'TensorFlow Edge TPU', 'TensorFlow.js', 'PaddlePaddle', 'MNN', 'NCNN', 'IMX', 'RKNN')
See https://docs.ultralytics.com/modes/predict for help.`

I believe 8001 is indeed the right port. May I share the entire script?
